config.org
#+startup: overview
#+title: Emacs Config File
#+author: KSH
#+email: skye@harith.xyz
#+options: toc:t todo:nil

* User Information
This contains basic user information regarding the user such as email id and name. It is used throughout emacs in places such as Org mode preambles and Mu4e.

These variables are defined in the core C code of emacs.

#+begin_src emacs-lisp
  ;; -*- lexical-binding: t -*-
  (setq user-full-name "Kaushik Srinivasan Harith")
  (setq user-mail-address "kaushik@archeanchemicals.com")
#+end_src

* Tweaks to emacs' functioning
There are some basic tweaks to make to emacs' default behaviour that make it more intuitive for use.
** COMMENT PATH variable
Sometimes certain packages in emacs make calls to executeables from the system. This means we occasionally need to add to this path variable here to let emacs be aware of those commands.

#+begin_src emacs-lisp
#+end_src

** Better autosaving
Auto-saving for default emacs puts the auto saved files in the current working directory. It makes more sense to prevent cluttering in organized folders by dumping these autosave files in a dedicated place.

Other changes include backups via file copying, giving priority to newer versions for backup storage, and automatic deleting of out-dated versions without prompting. Files are noteably version controled.

These variables are defined in =files.el=.

#+begin_src emacs-lisp
  (setq backup-directory-alist '(("." . "~/.emacs.d/saves/"))
        backup-by-copying t
        delete-old-versions t
        kept-new-versions 6
        kept-old-versions 2
        version-control t)

  (setq auto-save-file-name-transforms
        `((".*" "~/.emacs.d/saves/" t)))
#+end_src

** Window Management
This is a fix to make emacs windows place more consistently in the frame nicked from Protesilaous Stavrous. Basically, you use Regex to identify the type of buffer that spawns and place it accordingly in a special kind of window called a =side-window=. You can control the height and width of the side-window as a ratio of the whole screen. Note that when the side-window is the only window on the screen, you can't kill the frame.

This behaviour is controlled by =window.el=.
#+begin_src emacs-lisp
  (use-package window
    :ensure nil
    :init
    (setq display-buffer-alist
          '(;; top window
            ("\\*Messages.*"
             (display-buffer-in-side-window)
             (window-height . 0.25)
             (side . top)
             (slot . 1))
            ("\\*\\(Backtrace\\|Warnings\\|Compile-Log\\)\\*"
             (display-buffer-in-side-window)
             (window-height . 0.25)
             (side . top)
             (slot . 2))
            ;; bottom window
            ("^\\(\\*e?shell\\|vterm\\).*" ;; You don't use eshell. get rid of it
             (display-buffer-in-side-window)
             (window-width . 0.40)
             (side . bottom)
             (slot . 1))
            ;; left side window
            ("\\*Help.*"
             (display-buffer-in-side-window)
             (window-width . 0.35)       ; See the :hook
             (side . left)
             (slot . 0))
            ;; right window
            ("\\*Faces\\*"
             (display-buffer-in-side-window)
             (window-width . 0.35)
             (side . right)
             (slot . 0)
             (window-parameters . ((mode-line-format . (" " mode-line-buffer-identification)))))
            ("\\*Custom.*"
             (display-buffer-in-side-window)
             (window-width . 0.35)
             (side . right)
             (slot . 1))
            ))

    (setq window-combination-resize t)
    (setq even-window-sizes 'height-only)
    (setq window-sides-vertical nil)

    :hook ((help-mode . visual-line-mode)
           (custom-mode . visual-line-mode))
           )
#+end_src

** COMMENT Vertical splits as defualt
Modern monitors are wider than they are tall. Latin text also flows from top to bottom. It makes sense then to naturally divide a frame across the width of the screen instead of across the length.

This makes sense to me since i primarily use emacs full-screen. If you use a tiling wm or prefer narrower frames (say if you were using the =nano-emacs= configuration ), then go ahead and comment these out.

These variables are defined in =window.el=.

#+begin_src emacs-lisp
  (setq split-height-threshold nil)
  (setq split-width-threshold 0)
#+end_src

** Custom-set variables
When you customize a variable in emacs using the =M-x customize-variable= interface, it dumps all the customizations in the init.el file. This is not that big a deal, but to improve readability of these files, we can add a dedicated file for this.

This behaviour is defined in =cus-edit.el=.

#+begin_src emacs-lisp
  (setq custom-file (concat user-emacs-directory "custom.el"))

  (load-file custom-file)
#+end_src

** Editing changes
The following are some optional changes I make to emacs' editing experience.

*** COMMENT Cua Mode
I have grown accustomed to the church of emacs way of doing things. However, if one prefers, they can do things the 'normal' way by enabling =cua-mode= by uncommenting out this section.

This behaviour is defined in =cua-base.el=.

#+begin_src emacs-lisp
  (cua-mode t)
#+end_src

*** Replace selected text with typed text
By default, when you select a region in emacs and start typing, it doesn't replace the text as would be expected when migrating from more CUA standard apps like word. This changes that behavior

This behaviour is defined in =delsel.el=.

#+begin_src emacs-lisp
  (delete-selection-mode 1)
#+end_src

*** Visual Line Mode
By default emacs will let a line go off screen to infinity unless a =<return>= key is pressed to mark the end of the line. This behaviour made sense back in the days of type writers and carraige returns but makes no sense in the modern monitor based world.

The function is defined in =simple.el=.
#+begin_src emacs-lisp
    (global-visual-line-mode t)
#+end_src

** Custom functions
In the course of using emacs, I have found it useful to collect and define certain functions for simple tasks. Functions obtained from the community bear the prefix =contrib/=, while my own bear the prefix =skye/=. This list was a *lot longer* but I have decided to trim it down to the essentials.

*** Minibuffer quitting
Quitting out of the minibuffer can be a bit finicky. This function forces focus to the minibuffer when it is "active", thus making =C-g= quitting more reliable.

#+begin_src emacs-lisp
  (defun contrib/keyboard-quit-context+ ()
    "Quit current context.

  This function is a combination of `keyboard-quit' and
  `keyboard-escape-quit' with some parts omitted and some custom
  behavior added."
    (interactive)
    (cond ((region-active-p)
           ;; Avoid adding the region to the window selection.
           (setq saved-region-selection nil)
           (let (select-active-regions)
             (deactivate-mark)))
          ((eq last-command 'mode-exited) nil)
          (current-prefix-arg
           nil)
          (defining-kbd-macro
           (message
            (substitute-command-keys
             "Quit is ignored during macro defintion, use \\[kmacro-end-macro] if you want to stop macro definition"))
           (cancel-kbd-macro-events))
          ((active-minibuffer-window)
           (when (get-buffer-window "*Completions*")
             ;; hide completions first so point stays in active window when
             ;; outside the minibuffer
             (minibuffer-hide-completions))
           (abort-recursive-edit))
          (t
           (when completion-in-region-mode
             (completion-in-region-mode -1))
           (let ((debug-on-quit nil))
             (signal 'quit nil)))))
#+end_src

*** Quit Window
Useful for things like dired etc that don't kill the window on quitting.

#+begin_src emacs-lisp
  (defun skye/quit-window ()
    "If more than one window is open, close window on quit"
    (interactive)
    (if (> (length (window-list)) 1) (delete-window) (quit-window)))
#+end_src

*** Org PDF convenience tools
These are a bunch of nifty functions to reload and recompile latex files and pdf files fully from within emacs.

#+begin_src emacs-lisp
  (defun contrib/update-other-buffer ()
    (interactive)
    (other-window 1)
    (revert-buffer nil t)
    (other-window -1))

  (defun contrib/latex-compile-and-update-other-buffer ()
    "Has as a premise that it's run from a latex-mode buffer and the
     other buffer already has the PDF open"
    (interactive)
    (save-buffer)
    (shell-command (concat "pdflatex " (buffer-file-name)))
    (switch-to-buffer (other-buffer))
    (kill-buffer)
    (contrib/update-other-buffer))

  (defun contrib/org-compile-beamer-and-update-other-buffer ()
    "Has as a premise that it's run from an org-mode buffer and the
     other buffer already has the PDF open"
    (interactive)
    (org-beamer-export-to-pdf)
    (contrib/update-other-buffer))

  (defun contrib/org-compile-latex-and-update-other-buffer ()
    "Has as a premise that it's run from an org-mode buffer and the
     other buffer already has the PDF open"
    (interactive)
    (org-latex-export-to-pdf)
    (contrib/update-other-buffer))

  (eval-after-load 'latex-mode
    '(define-key latex-mode-map (kbd "C-c r") 'contrib/latex-compile-and-update-other-buffer))

  (define-key org-mode-map (kbd "C-c lr") 'contrib/org-compile-latex-and-update-other-buffer)
  (define-key org-mode-map (kbd "C-c br") 'contrib/org-compile-beamer-and-update-other-buffer)
#+end_src

** Custom Keybinds
This is a list of custom keybinds that I couldn't fit in with the package declarations. Use =C-z= for all my custom keybinds and use =M-z= for functions that are related but broader/inverse in scope.

#+begin_src emacs-lisp
  (use-package emacs
    :ensure nil
    :bind
    (:map global-map
          :prefix-map ctrl-z-prefix-map
          :prefix "C-z"
          ("C-<SPC>" . fixup-whitespace)
          ("C-e" . eval-defun)
          (";" . comment-dwim)
          ("C-h f" . describe-face))

    (:map global-map
          ("<f5>" . revert-buffer)
          ([remap kill-buffer] . skye/kill-current-buffer)
          )
    ) 
#+end_src

* Package.el and Use-package setup
The following lines of code initialize the file =package.el=, then add to a list - variable provided by =package.el= called =package-archives= which lets it know which archives (lists of repositories of packages) to use.

By default, GNU ELPA (standing for Emacs Lisp Package Archive) and Non-GNU ELPA are the values in the list.

Once that is done, we install a handy package called =vc-use-package= that works in conjunction with =use-package= (now built into emacs) to handle installing packages from other sources such as github, gitlab etc. The use-package macro is used extensively in this configuration to improve the readability to the user and to reduce unnecesary complexity.

We lastly tell =use-package= to install a package if it is not available with the variable =use-package-always-ensure=.

Use-package is defined in =use-package-core.el=.

#+begin_src emacs-lisp
  (require 'package)
  (add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)

  (unless (package-installed-p 'vc-use-package)
    (package-vc-install "https://github.com/slotThe/vc-use-package"))

  (setq use-package-always-ensure t)
#+end_src

* Tweaks to emacs' appearance
The complete look and feel of emacs is being dictated by nano emacs. It is objectively the best looking emacs distribution. I Just modify the colors set by =nano-base-colors.el= to match my own emacs theme, TransSide

#+begin_src emacs-lisp
  (defun skye/nano-dark-theme ()
    "Modifying the nano theme with my own colors (dark)"
    (setq nano-color-background "#00000F"
          nano-color-critical "#FF0069"
          nano-color-faded "#BE9DDF"
          nano-color-foreground "#FFEEFF"
          nano-color-highlight "#AEC6FF"
          nano-color-popout "#FEA5D8"
          nano-color-salient "#92DDFA"
          nano-color-strong "#EFDFEF"
          nano-color-subtle "#6E58BB"
          nano-theme-var "dark")
    (nano-refresh-theme)
  )

  (defun skye/nano-light-theme ()
    "Modifying the nano theme with my own colors (light)"
    (setq nano-color-background "#F1EBFF"
          nano-color-critical "#FF0069"
          nano-color-faded "#504099"
          nano-color-foreground "#1F001F"
          nano-color-highlight "#EFCFEF"
          nano-color-popout "#800037"
          nano-color-salient "#0068AD"
          nano-color-strong "#000000"
          nano-color-subtle "#816ECF"
          nano-theme-var "light")
    (nano-refresh-theme)
  )

  (defun skye/nano-catppuccin-mocha ()
    "Modifying the nano theme with my own colors (dark)"
    (setq nano-color-background "#1E1E2E"
          nano-color-critical "#F38BA8"
          nano-color-faded "#BAC2DE"
          nano-color-foreground "#EDF6FF"
          nano-color-highlight "#89DCFB"
          nano-color-popout "#8F71E7"
          nano-color-salient "#54CFFF"
          nano-color-strong "#EDF6FF"
          nano-color-subtle "#313244"
          nano-theme-var "dark")
    (nano-refresh-theme)
    )

  (use-package nano
    :vc (:fetcher github :repo rougier/nano-emacs)
    :config
    (nano-theme-set-dark)
    (skye/nano-catppuccin-mocha)
    )

  (set-face-attribute 'default nil :weight 'regular)
#+end_src

* Packages
** Packages that make navigating emacs easier
*** All the icons 
All-the-icons provides fancy icons for all the lists and menus in emacs. This code uses the 6.0 beta which bundles the svgs instead of downloading the fonts.

#+begin_src emacs-lisp
  (use-package all-the-icons
    :vc (all-the-icons :url "https://github.com/domtronn/all-the-icons.el.git" :branch "svg")
    :if (display-graphic-p))

  (use-package all-the-icons-completion)

  (use-package all-the-icons-dired)
#+end_src

*** cdLaTeX
=cdlatex= is a useful package when typesetting math in LaTeX. Pressing the ~ key activates an overlay for you to choose from the most common math and logic symbols.

#+begin_src emacs-lisp
  (use-package cdlatex)
#+end_src

*** Better minibuffer
A lot happens in the emacs minibuffer. Text search (default by =isearch, ido=), which inturn requires text completion (default by =icomplete=), the help selection for functions, variables etc, and a lot more.

The following group of packages =vertico, marginalia, consult, orderless= are made by one Minad on github. They are very close to vanilla emacs but provide a richer experience with a focus on a small codebase.

#+begin_src emacs-lisp
  (use-package vertico
    :bind
    (:map vertico-map
          ([remap keyboard-quit] . contrib/keyboard-quit-context+))
    :custom
    (vertico-count 8)
    (vertico-resize t)
    (vertico-cycle nil)
    :init
    (vertico-mode)
    )

  (use-package marginalia
    :custom
    (marginalia-max-relative-age 0)
    (marginalia-align 'right)
    :init
    (marginalia-mode)
    :hook
    (marginalia-mode . all-the-icons-completion-marginalia-setup))

  (all-the-icons-completion-mode)
  
  (use-package consult
    :bind
    (:map global-map
          ("C-s" . consult-line)
          ("C-x b" . consult-buffer)
          ))

  (use-package orderless
    :custom

config.el
(setq user-full-name "Kaushik Srinivasan Harith")
(setq user-mail-address "kaushik.harith@gmail.com")

(use-package all-the-icons
  :ensure t
  :diminish "")

(use-package all-the-icons-dired
  :ensure t
  :diminish "")

(use-package all-the-icons-ibuffer
  :ensure t)

(all-the-icons-ivy-setup)
(all-the-icons-ivy-rich-mode 1)

(use-package prescient
  :ensure
  :config
  (setq prescient-history-length 200)
  (setq prescient-save-file "~/.emacs.d/prescient-items")
  (setq prescient-filter-method '(literal regexp))
  (prescient-persist-mode 1))

(use-package ivy-prescient
  :ensure
  :after (prescient ivy)
  :config
  (setq ivy-prescient-sort-commands
        '(:not counsel-grep
               counsel-rg
               counsel-switch-buffer
               ivy-switch-buffer
               swiper
               swiper-multi))
  (setq ivy-prescient-retain-classic-highlighting t)
  (setq ivy-prescient-enable-filtering nil)
  (setq ivy-prescient-enable-sorting t)
  (ivy-prescient-mode 1))

(use-package company
  :diminish ""
  :ensure t
  :init
  (setq company-require-match nil) ; Don't require match, so you can still move your cursor as expected.
  (setq company-tooltip-align-annotations t) ; Align annotation to the right side.
  (setq company-eclim-auto-save nil) ; Stop eclim auto save.
  (setq company-dabbrev-downcase nil) ; No downcase when completion.

  :config
  (setq company-idle-delay 0)
  (setq company-minimum-prefix-length 4)
  (add-hook 'prog-mode-hook 'company-mode)

  (defun jcs--company-complete-selection--advice-around (fn)
    "Advice execute around `company-complete-selection' command."
    (let ((company-dabbrev-downcase t)) (call-interactively fn))) (advice-add 'company-complete-selection :around #'jcs--company-complete-selection--advice-around))

(use-package company-fuzzy
  :ensure t
  :after (company)
  :diminish ""
  :config
  (global-company-fuzzy-mode 1)
  (setq company-fuzzy-prefix-ontop t)
  (setq company-fuzzy-sorting-backend 'alphabetic)
  (setq company-fuzzy-show-annotation t))

(use-package company-auctex
  :ensure t
  :init
  (company-auctex-init))

(use-package counsel
  :ensure t)

(use-package swiper
  :ensure t)

(use-package ivy
  :demand
  :diminish ""
  :ensure t
  :bind
  (:map global-map
        ("C-s" . swiper)
        ("C-r" . swiper-backward)
        ("C-c C-r" . ivy-resume)
        ("<f6>" . ivy-resume)
        ("M-x" . counsel-M-x)
        ("C-x C-f" . counsel-find-file)
        ("s-f" . counsel-find-file)	
        ("<f1> f" . counsel-describe-function)
        ("<f1> v" . counsel-describe-variable)
        ("<f1> l" . counsel-load-library)
        ("<f2> i" . counsel-info-lookup-symbol)
        ("<f2> u" . counsel-unicode-char)
        ("C-c g" . counsel-git)
        ("C-c j" . counsel-git-grep)
        ("C-c k" . counsel-ag)
        ("C-x l" . counsel-locate)
        ("M-y" . counsel-yank-pop))
  :config
  (ivy-mode 1)
  (setq ivy-use-virtual-buffers t)
  (setq swiper-use-visual-line nil)
  (setq swiper-use-visual-line-p (lambda (a) nil)))

(use-package diminish
  :ensure t)

(use-package dired
  :config
  (setq dired-recursive-copies 'always)
  (setq dired-recursive-deletes 'always)
  (setq delete-by-moving-to-trash t)
  (setq dired-listing-switches "-al --group-directories-first --time-style=iso")
  (setq dired-dwim-target t)
  :hook
  ((dired-mode . dired-hide-details-mode)
   (dired-mode . all-the-icons-dired-mode))
  :bind
  (:map dired-mode-map ("q" . skye/quit-window)))

(electric-pair-mode 1)

(use-package expand-region
  :ensure t
  :after (org)
  :bind
  (:map global-map
        ("C-=" . er/expand-region)))

(use-package ibuffer
  :ensure t

  :bind
  (("C-x C-b" . ibuffer)
   ("s-b" . ibuffer))

  :config
  (setq ibuffer-expert t)
  (setq ibuffer-saved-filter-groups
        '(("home"
           ("Magit" (or (mode . magit-process-mode)
                        (mode . magit-diff-mode)
                        (mode . magit-mode)
                        (mode . magit)
                        (mode . magit-blame-mode)
                        (mode . magit-blob-mode)
                        (mode . magit-cherry-mode)
                        (mode . magit-file-mode)
                        (mode . magit-wip-initial-backup-mode)
                        (mode . magit-log-mode)
                        (mode . magit-log-select-mode)
                        (mode . magit-submodule-list-mode)))
           ("Latex" (or (mode . latex-mode)
                        (mode . bibtex-mode)
                        (mode . latex-mode)))
           ("Org" (mode . org-mode))
           ("Help" (or (name . "\*Help\*")
                       (name . "\*Apropos\*")
                       (name . "\*info\*")
                       (mode . special-mode)
                       (mode . messages-buffer-mode)
                       (mode . fundamental-mode))))))

  (add-hook 'ibuffer-mode-hook
            '(lambda ()
               (ibuffer-auto-mode 1)
               (ibuffer-switch-to-saved-filter-groups "home")
               (all-the-icons-ibuffer-mode 1)))

  )

(ivy-rich-mode 1)
(setcdr (assq t ivy-format-functions-alist) #'ivy-format-function-line)

(use-package lorem-ipsum
  :ensure t)

(use-package magit
  :ensure t)

(use-package ace-window
  :ensure t
  :bind
  ([remap other-window] . ace-window)
  :config
  (setq aw-keys '(?a ?s ?d ?f ?g ?h ?j ?k ?l))
  )

(use-package avy
  :ensure t
  :bind
  (:map global-map
        ("C-;" . avy-goto-char)
        ("C-'" . avy-goto-char-2)
        ("M-g l" . avy-goto-line)
        ("M-g r" . avy-resume)
        :map org-mode-map
        ("C-'" . avy-goto-char-2))
  )

(use-package rainbow-delimiters
  :ensure t
  :hook
  (prog-mode . rainbow-delimiters-mode)
  (Latex-mode . rainbow-delimiters-mode)
  :config
  (rainbow-delimiters-mode 1))

(use-package rainbow-mode
  :ensure t)

(use-package linum-relative
  :ensure t
  :init
  (global-linum-mode t)
  :config
  (linum-relative-mode)
  (add-hook 'doc-view-mode-hook 'contrib/inhibit-global-linum-mode))

(use-package restart-emacs
  :ensure t
  :config
  (setq restart-emacs-restore-frames t))

(use-package try
  :ensure t)

(use-package undo-tree
  :ensure t
  :diminish ""
  :init
  (global-undo-tree-mode)
  :config
  (setq undo-tree-history-directory-alist '(("." . "~/.emacs.d/undo"))))

(use-package which-key
  :diminish ""
  :ensure t
  :config (which-key-mode))

(use-package yasnippet
  :ensure t
  :config
  (yas-global-mode 1)

  (setq yas-snippet-dirs '("~/.emacs.d/snippets/")))

(use-package yasnippet-snippets
  :ensure t)

(use-package latex
  :defer t
  :ensure auctex
  :mode ("//.tex//" . latex-mode)
  :hook
  (LaTeX-mode . outline-minor-mode)
  :config
  (progn
    (setq TeX-fold-mode t)
    (setq TeX-parse-self t)
    (setq TeX-save-query nil)
    (setq TeX-PDF-mode t)
    (add-hook 'LaTeX-mode-hook 'cdlatex-mode)
    ))

(setenv "PATH" (concat "/opt/texlive/2020/bin/x86_64-linux:"
                       (getenv "PATH")))
(add-to-list 'exec-path "/opt/texlive/2020/bin/x86_64-linux")

;; (load "preview-latex.el" nil t t)

(use-package xenops
  :ensure t
  :hook
  (latex-mode . xenops-mode)
  (LaTeX-mode . xenops-mode))

(add-to-list 'load-path "~/.emacs.d/mu4e")

(require 'mu4e)
(require 'mu4e-contrib)

(setq mu4e-trash-folder "/TRASH")
(setq mu4e-sent-folder "/SENT")
(setq mu4e-drafts-folder "/drafts")

(setq mu4e-headers-results-limit 500)

;; tell message-mode how to send mail
(setq message-send-mail-function 'smtpmail-send-it)
;; if our mail server lives at smtp.example.org; if you have a local
;; mail-server, simply use 'localhost' here.
(setq smtpmail-smtp-server "smtp.gmail.com")
(setq smtpmail-smtp-service 587)       
(setq smtpmail-stream-type 'starttls)

(setq mu4e-get-mail-command "mbsync -a")

(use-package org
  :ensure org-superstar
  :hook
  (org-mode . org-cdlatex-mode)
  (org-mode . (lambda () (org-superstar-mode)))
  (org-mode . org-indent-mode)
  :init
  (setq org-highlight-latex-and-related '(native latex script))
  (setq org-export-backends '(ascii html icalendar latex odt org))
  :bind
  (:map org-mode-map
        ("C-c C-x C-e" . skye/org-mark-and-archive)
        ("C-c C-x <up>" . org-cycle-list-bullet)
        :map global-map
        ("C-c a" . org-agenda)
        ("C-c c" . org-capture))
  :config
  (require 'org-tempo)

  (setq org-directory "~/Documents/life/")

  (defun skye/org-get-path (stringname)
    "Use concat to generate full path."
    (concat (file-name-as-directory org-directory) stringname))

  (setq skye/Readme (skye/org-get-path "README.org"))
  (setq skye/Ideas (skye/org-get-path "Ideas.org"))
  (setq skye/School (skye/org-get-path "SchoolWork.org"))
  (setq skye/archive (skye/org-get-path "archive.org"))
  (setq skye/calendar-personal (skye/org-get-path "calendar-personal.org"))
  (setq skye/calendar-stony (skye/org-get-path "calendar-stony.org"))

  (setq org-agenda-files (list skye/Readme skye/Ideas skye/School))
  (setq org-archive-location (concat skye/archive "::* From %s"))

  (setq org-ellipsis " ▼")
  (setq org-src-fontify-natively t)
  (setq org-src-tab-acts-natively t)

  (setq org-todo-keywords '((sequence "☛TODO(t)" "|" "⚑WAITING(w!)") (sequence "|" "❌CANCELED(c)" "|" "✔DONE(d)")))

  (setq org-enforce-todo-dependencies t)
  (setq org-enforce-todo-checkbox-dependencies t)

  (setq skye/bullets-list '("◉" "●" "○" "⊙"))

  (setq org-src-window-setup 'current-window)

  (defun skye/org-mark-and-archive ()
    "Mark the state of the current subtree as either DONE or CANCELLED and export to my archive.org file"
    (interactive)
    (ivy-read "Choose a final TODO state:" '("✔DONE" "❌CANCELED")
              :action '(1
                        ("o" org-todo "action 1")
                        ("j" org-todo "action 2")))
    (org-archive-subtree))

  (add-to-list 'org-modules 'org-habit)

  (setq org-habit-graph-column 80)
  (setq org-habit-show-habits-only-for-today t)
  (setq org-habit-show-all-today t)

  (setq org-default-notes-file skye/Readme)

  (setq org-capture-templates '(
                                ("e" "Email to be dealt with. Action item" entry
                                 (file+headline skye/Readme "Emails to deal with")
                                 "* ☛TODO %:from %? \n %a \n SCHEDULED: %^t DEADLINE: %^t \n :PROPERTIES: \n CREATED: %u \n :END:"
                                 )

                                ("m" "Miscellaneous TODO. Refile" entry
                                 (file+headline skye/Readme "Miscellaneous")
                                 "* ☛TODO %^{PROMPT} %? \n SCHEDULED: %^t DEADLINE: %^u"
                                 )

                                ("t" "Date-less TODO. Generic" entry
                                 (file skye/Readme)
                                 "* ☛TODO %^{PROMPT} \n  %?"
                                 )
                                ))

  (setq org-refile-targets
        '((nil :maxlevel . 3)
          (org-agenda-files :maxlevel . 2)))

  (setq org-pretty-entities nil)
  (setq org-preview-latex-default-process 'dvisvgm)
  )

(use-package org-superstar
  :ensure t
  :after
  (org)
  :config
  (setq org-superstar-leading-bullet ?\s)
  (setq org-superstar-cycle-headline-bullets t)
  (setq org-superstar-headline-bullets-list skye/bullets-list))

(setq inhibit-startup-message t) ;startup.el
(tool-bar-mode -1) ;tool-bar.el
(menu-bar-mode -1) ;menu-bar.el
(set-scroll-bar-mode nil) ;sroll-bar.el
(global-visual-line-mode t) ;simple.el
(set-fringe-mode 0) ;fringe.el
(tab-bar-mode 0) ;tab-bar.el
(global-tab-line-mode 0) ;tab-line.el

(set-face-attribute 'default nil :font "Inconsolata LGC Markup" :height 130)

(add-to-list 'custom-theme-load-path "~/.emacs.d/themes/")
(load-theme 'TransSide t)

(add-to-list 'default-frame-alist '(fullscreen . fullboth))

(use-package nano-modeline
  :ensure t
  :config
  (setq nano-modeline-position 'mode-line)
  (nano-modeline-mode))

(fset 'yes-or-no-p 'y-or-n-p)

(setq backup-by-copying t      ; don't clobber symlinks
      backup-directory-alist '(("." . "~/.emacs.d/saves/"))    ; don't litter my fs tree
      delete-old-versions t
      kept-new-versions 6
      kept-old-versions 2
      version-control t)       ; use versioned backups

(setq auto-save-file-name-transforms
      `((".*" "~/.emacs.d/saves/" t)))

(setq split-height-threshold nil) ;window.el
(setq split-width-threshold 0) ;window.el

(use-package window
  :init
  (setq display-buffer-alist
        '(;; top side window
          ("\\*\\(Flycheck\\|Flymake\\|Package-Lint\\|vc-git :\\).*" ;; This bit is useless to me currently. Rethink it later. 
           (display-buffer-in-side-window)
           (window-height . 0.25)
           (side . top)
           (slot . 0))
          ("\\*Messages.*"
           (display-buffer-in-side-window)
           (window-height . 0.25)
           (side . top)
           (slot . 1))
          ("\\*\\(Backtrace\\|Warnings\\|Compile-Log\\)\\*"
           (display-buffer-in-side-window)
           (window-height . 0.25)
           (side . top)
           (slot . 2))
          ;; bottom side window
          ("\\*\\(Output\\|Register Preview\\).*"
           (display-buffer-in-side-window)
           (window-width . 0.20)       ; See the :hook
           (side . bottom)
           (slot . -1))
          (".*\\*\\(Completions\\|Embark.*Occur\\).*"
           (display-buffer-in-side-window)
           (window-height . 0.25)
           (side . bottom)
           (slot . 0))
          ("^\\(\\*e?shell\\|vterm\\).*" ;; You don't use eshell. get rid of it
           (display-buffer-in-side-window)
           (window-width . 0.40)
           (side . right)
           (slot . 1))
          ;; left side window
          ("\\*Help.*"
           (display-buffer-in-side-window)
           (window-width . 0.25)       ; See the :hook
           (side . left)
           (slot . 0))
          ;; right side window
          ("\\*Faces\\*"
           (display-buffer-in-side-window)
           (window-width . 0.25)
           (side . right)
           (slot . 0)
           (window-parameters . ((mode-line-format . (" " mode-line-buffer-identification)))))
          ("\\*Custom.*"
           (display-buffer-in-side-window)
           (window-width . 0.25)
           (side . right)
           (slot . 1))
          ;; bottom buffer (NOT side window)
          ("\\*\\vc-\\(incoming\\|outgoing\\).*"
           (display-buffer-at-bottom))))
  (setq window-combination-resize t)
  (setq even-window-sizes 'height-only)
  (setq window-sides-vertical nil)
  ;; Note that the the syntax for `use-package' hooks is controlled by
  ;; the `use-package-hook-name-suffix' variable.  The "-hook" suffix is
  ;; not an error of mine.
  :hook ((help-mode . visual-line-mode)
         (custom-mode . visual-line-mode))
  :bind (("s-n" . next-buffer)
         ("s-p" . previous-buffer)
         ("s-o" . other-window)
         ("s-0" . delete-window)
         ("s-1" . delete-other-windows)
         ("s-5" . delete-frame)
         ("C-x +" . balance-windows-area)))

(setq auto-window-vscroll nil)

(setq custom-file (concat user-emacs-directory "custom.el"))

(load-file custom-file)

(prefer-coding-system       'utf-8)
(set-default-coding-systems 'utf-8)
(set-terminal-coding-system 'utf-8)
(set-keyboard-coding-system 'utf-8)
(setq default-buffer-file-coding-system 'utf-8)
(setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING))

(use-package unicode-fonts
  :ensure t
  :config
  (unicode-fonts-setup))

(defun contrib/toggle-window-split ()
  (interactive)
  (if (= (count-windows) 2)
      (let* ((this-win-buffer (window-buffer))
             (next-win-buffer (window-buffer (next-window)))
             (this-win-edges (window-edges (selected-window)))
             (next-win-edges (window-edges (next-window)))
             (this-win-2nd (not (and (<= (car this-win-edges)
                                         (car next-win-edges))
                                     (<= (cadr this-win-edges)
                                         (cadr next-win-edges)))))
             (splitter
              (if (= (car this-win-edges)
                     (car (window-edges (next-window))))
                  'split-window-horizontally
                'split-window-vertically)))
        (delete-other-windows)
        (let ((first-win (selected-window)))
          (funcall splitter)
          (if this-win-2nd (other-window 1))
          (set-window-buffer (selected-window) this-win-buffer)
          (set-window-buffer (next-window) next-win-buffer)
          (select-window first-win)
          (if this-win-2nd (other-window 1))))))

(defun contrib/keyboard-quit-context+ ()
  "Quit current context.

This function is a combination of `keyboard-quit' and
`keyboard-escape-quit' with some parts omitted and some custom
behavior added."
  (interactive)
  (cond ((region-active-p)
         ;; Avoid adding the region to the window selection.
         (setq saved-region-selection nil)
         (let (select-active-regions)
           (deactivate-mark)))
        ((eq last-command 'mode-exited) nil)
        (current-prefix-arg
         nil)
        (defining-kbd-macro
          (message
           (substitute-command-keys
            "Quit is ignored during macro defintion, use \\[kmacro-end-macro] if you want to stop macro definition"))
          (cancel-kbd-macro-events))
        ((active-minibuffer-window)
         (when (get-buffer-window "*Completions*")
           ;; hide completions first so point stays in active window when
           ;; outside the minibuffer
           (minibuffer-hide-completions))
         (abort-recursive-edit))
        (t
         (when completion-in-region-mode
           (completion-in-region-mode -1))
         (let ((debug-on-quit nil))
           (signal 'quit nil)))))

(global-set-key [remap keyboard-quit] #'contrib/keyboard-quit-context+)

(defun contrib/inhibit-global-linum-mode ()
  "Counter-act `global-linum-mode'."
  (add-hook 'after-change-major-mode-hook (lambda () (linum-mode 0)) :append :local))

(defun skye/kill-current-buffer ()
  (interactive)
  (kill-buffer (current-buffer)))

(defun skye/small-text-scale ()
  "Make the buffer text zoom smaller"
  (interactive)
  (text-scale-set -1))

(defun skye/big-text-scale ()
  "Make the buffer text zoom bigger"
  (interactive)
  (text-scale-set 1))

(defun skye/quit-window ()
  "If more than one window is open, close window on quit"
  (interactive)
  (if (> (length (window-list)) 1) (delete-window) (quit-window)))

(defun skye/reload-theme ()
  "Reload a theme by disabling that theme first"
  (interactive)
  (let ((theme 'TransSide))
  (eval-buffer (current-buffer))
  (disable-theme theme)
  (enable-theme theme)))

(use-package emacs
  :bind
  (:map global-map
        :prefix-map my-ctrl-z-prefix-map
        :prefix "C-z"
        ("C-<SPC>" . fixup-whitespace)
        ("C-e" . eval-defun)
        ("|" . contrib/toggle-window-split)
        (";" . comment-region)
        ("C-h f" . describe-face))

  (:map global-map
        :prefix-map my-meta-z-prefix-map
        :prefix "M-z"
        (";" . uncomment-region))

  (:map global-map
        ("<f5>" . revert-buffer)
        ([remap kill-buffer] . skye/kill-current-buffer)
        ("s-s" . save-buffer)
        ("<f4>" . skye/reload-theme)
        )
  )

(put 'scroll-left 'disabled nil)
(put 'dired-find-alternate-file 'disabled nil)
(put 'narrow-to-region 'disabled nil)
